Selection scraper created and integrated to project

kursat opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/5142

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/5142

### Tests & QA Steps
<!--- 
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

#### Android / IOS
1. Send some messages contains markdown.
2. Copy with context menu.
3. Ensure copied message is same with sent message.

#### Web / Mobile Web / Desktop
1. Send some messages contains markdown.
2. Copy with context menu.
3. Copy all message by selecting with mouse.
4. Copy single line of message by selecting with mouse/double clicking.
5. Copy a word of message by selecting with mouse/double clicking.
6. Copy multiple line of message by selecting with mouse.
6. Copy by selecting multiple messages with single/multiple lines. (This will not work as intended until another bug is fixed)
7. Ensure copied message is same with sent message.

#### Markdowns to Test
1. Code Block
2. Inline Code Block
3. Email
4. Link
5. Italic
6. Bold
7. Strikethrough
8. Quote
9. New line

#### Test String I've Used
```md
    This is `code`
    ```
    this is block code
    ```
    Mail to: [Expensify](concierge@expensify.com)
    Link: [Expensify](https://www.expensify.com)
    _italic_

    *bold*
    ~strikethrough~
    > Quote
```

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

I'll upload IOS and Desktop videos ASAP.

#### Web
<!-- Insert screenshots of your changes on the web platform-->
https://user-images.githubusercontent.com/1279379/145830278-340c8d80-a6ee-4e6d-885a-60b8d82622a9.mp4

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
https://user-images.githubusercontent.com/1279379/145829802-0c7f61be-a580-47df-88db-1a6bc5f826a9.mp4

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
https://user-images.githubusercontent.com/1279379/146242081-245546f5-98e7-4ce8-aafb-157df43d7086.mp4

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
https://user-images.githubusercontent.com/1279379/146216218-1b18dd9c-fc51-4c20-bcf8-f22dbd0223a8.mp4

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
https://user-images.githubusercontent.com/1279379/145832122-78e845c6-7fce-4fe7-b63d-c3c214a57e7a.mp4

kursat committed:
Selection scraper created and integrated to project

thienlnam commented on src/libs/SelectionScraper/index.js:
```diff
@@ -0,0 +1,115 @@
+const tagAttribute = 'data-testid';
+
+/**
+ * Reads html of selection. If there is no Selection API returns empty string.
```
If there is no selection, the API

kursat replied:
I meant if browser doesn't support [Selection API](https://caniuse.com/selection-api), Is it OK if I update it like that?

kursat replied:
Hi @thienlnam, I've updated all comments, can you recheck, if they are OK?

thienlnam replied:
Yes that update is good 👍  (just missing one more comment I left unresolved)

thienlnam commented on src/libs/SelectionScraper/index.js:
```diff
@@ -0,0 +1,115 @@
+        if (selection.rangeCount > 0) {
+            const div = document.createElement('div');
+
+            // We traverse all ranges, and get closest node with data-testid and replace its contents with contents of
+            // range.
```
Could we get a comment here on why we're looking for the node with data-testid?

thienlnam commented on src/libs/SelectionScraper/index.js:
```diff
@@ -0,0 +1,115 @@
+
+                    if (range.commonAncestorContainer instanceof HTMLElement) {
+                        node = range.commonAncestorContainer.closest(`[${tagAttribute}]`);
+                    } else {
+                        node = range.commonAncestorContainer.parentNode.closest(`[${tagAttribute}]`);
+                    }
```
When is this the case? / What does this else indicate

thienlnam replied:
Still looking for this comment, on why the parentNode would have the closest tag attribute instead of the current node

thienlnam commented on src/libs/SelectionScraper/index.js:
```diff
@@ -0,0 +1,115 @@
+                    }
+
+                    // This means "range.commonAncestorContainer" is a text node. We simply get its parent node.
+                    if (!node) { node = range.commonAncestorContainer.parentNode; }
```
Let's break this if statement into separate lines for consistency

thienlnam commented on src/libs/SelectionScraper/index.js:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    // If no Selection API returns empty string.
+    return '';
```
Same comment as above, If no selection, API returns empty string - or maybe I'm just reading it incorrectly

thienlnam commented on src/libs/SelectionScraper/index.js:
```diff
@@ -0,0 +1,115 @@
+    // We are skipping elements which has html and body in data-testid, since ExpensiMark can't parse it. Also this data
+    // has no meaning for us.
+    if (dom.attribs && dom.attribs[tagAttribute]) {
+        if (!elementsWillBeSkipped.includes(dom.attribs[tagAttribute])) { domName = dom.attribs[tagAttribute]; }
```
Break if statement into new lines

thienlnam reviewed (changes requested):
Looks solid! Just a few formatting comments

kursat committed:
comment update / code style

kursat committed:
fixed package versions

thienlnam reviewed (changes requested):
Comments look good, missing one comment and also looks like you have a large change with your package-lock.json that should be removed

kursat committed:
Revert "fixed package versions"

This reverts commit b7895a84c7ddd65515b4a893ddc08733e31265db.

kursat committed:
removed package-lock.json changes, re-committing fixed versions.

kursat committed:
Fixed a bug which causes no newlines after comments

kursat committed:
comment on selecting parentNode.closest()

kursat commented:
Hey @thienlnam, sorry for the package-lock.json file, I removed it but now all tests are failing. Is this normal? Also added the comment you requested along with a bugfix for Chrome.

thienlnam commented:
Sorry I should have clarified, package-lock.json file should be included but for some reason your package-lock.json diff seemed to be way too large to just include the new changes for packages you installed.

The tests are failing since you added new packages in package.json but the package-lock.json doesn't have them.

Could you try doing an npm install and then adding those changes again?

kursat committed:
added package-lock.json (installed dependencies with lts/erbium)

kursat commented:
@thienlnam, I updated package-lock.json file. Can you check please?

 It looks like using newer version of node.js or npm causes big changes on package-lock.json file.

```bash
kursat@kursat-laptop:~$ node -v
v14.18.1
kursat@kursat-laptop:~$ npm -v
8.2.0
kursat@kursat-laptop:~/WebstormProjects/upwork/ExpensifyApp$ npm i
npm WARN old lockfile 
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
```
Can we add an engines section to package.json file to prevent other contributors to face the same problem?

thienlnam commented:
Yup that looks much better, thanks! 

> Can we add an engines section to package.json file to prevent other contributors to face the same problem?

This is definitely something we can explore - feel free to propose the issue in the #expensify-open-source slack channel and then we can get the ball rolling from there

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1651672957) to staging by @thienlnam in version: 1.1.24-19 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|failure ❌
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|failure ❌

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1651839606) to production by @francoisl in version: 1.1.24-19 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|failure ❌
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|failure ❌
